src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/config/SchedulerConfig.java

```diff
@@ -1,21 +1,129 @@
+
+    @Bean
+    public TaskScheduler orchestratorTaskScheduler() {
+        ThreadPoolTaskScheduler scheduler = new OrchestratorTaskScheduler();
         scheduler.setPoolSize(POOL_SIZE);
+        scheduler.setThreadNamePrefix("BSO-");
+        scheduler.setErrorHandler(t -> {
+            log.error("Unhandled exception during task. {}: {}", t.getClass(), t.getMessage(), t);
+            errorCount.incrementAndGet();
+        });
         scheduler.initialize();
 
-        taskRegistrar.setTaskScheduler(scheduler);
+        return scheduler;
+    }
+
+    /**
+     * Custom {@link ThreadPoolTaskScheduler} to be able to register scheduled tasks via AppInsights.
+     */
+    private static class OrchestratorTaskScheduler extends ThreadPoolTaskScheduler {
+
+        private static final long serialVersionUID = 8789551087577035719L;
+
+        @Override
+        public void execute(Runnable command) {
+            super.execute(new WrappedRunnable(command, REQUEST_CONTEXT_SUPPLIER.get()));
+        }
+
+        @Override
+        public ScheduledFuture<?> schedule(Runnable task, Trigger trigger) {
+            return super.schedule(new WrappedRunnable(task, REQUEST_CONTEXT_SUPPLIER.get()), trigger);
+        }
+
+        @Override
+        public ScheduledFuture<?> schedule(Runnable task, Date startTime) {
+            return super.schedule(new WrappedRunnable(task, REQUEST_CONTEXT_SUPPLIER.get()), startTime);
+        }
+
+        @Override
+        public ScheduledFuture<?> scheduleAtFixedRate(Runnable task, Date startTime, long period) {
+            return super.scheduleAtFixedRate(new WrappedRunnable(
+                task,
+                REQUEST_CONTEXT_SUPPLIER.get()
+            ), startTime, period);
+        }
+
+        @Override
+        public ScheduledFuture<?> scheduleAtFixedRate(Runnable task, long period) {
+            return super.scheduleAtFixedRate(new WrappedRunnable(task, REQUEST_CONTEXT_SUPPLIER.get()), period);
+        }
+
+        @Override
+        public ScheduledFuture<?> scheduleWithFixedDelay(Runnable task, Date startTime, long delay) {
+            return super.scheduleWithFixedDelay(new WrappedRunnable(
+                task,
+                REQUEST_CONTEXT_SUPPLIER.get()
+            ), startTime, delay);
+        }
+
+        @Override
+        public ScheduledFuture<?> scheduleWithFixedDelay(Runnable task, long delay) {
+            return super.scheduleWithFixedDelay(new WrappedRunnable(task, REQUEST_CONTEXT_SUPPLIER.get()), delay);
+        }
+    }
+
+    private static class WrappedRunnable implements Runnable {
+
+        private final Runnable task;
+        private RequestTelemetryContext requestContext;
+
+        WrappedRunnable(Runnable task, RequestTelemetryContext requestContext) {
+            this.task = task;
+            this.requestContext = requestContext;
+        }
+
+        @Override
+        public void run() {
+            if (ThreadContext.getRequestTelemetryContext() != null) {
```

Comment:
What if '== null'?

Reply:
`requestContext` will be just set as in line 124.

Reply:
Ah, I see

Reply:
'requestContext' class attribute does not look necessary, it is refreshed each time and set to ThreadContext. It looks it is possible just do this:

RequestTelemetryContext requestContext = REQUEST_CONTEXT_SUPPLIER.get();
ThreadContext.setRequestTelemetryContext(requestContext);

Reply:
This is following Microsoft's application insights guidelines on how to track schedules correctly: https://github.com/microsoft/ApplicationInsights-Java/wiki/Distributed-Tracing-in-Asynchronous-Java-Applications#context-propagation-in-scheduled-events.

I believe `ThreadContext` can be cleared at any point in the request lifetime. Most likely once it's published via telemetry client. Then new registered context will be used upon next schedule run making this `if` skipped

Reply:
I see, this is just how Microsoft recommends